controllers/reconcilers/kserve_raw_route_reconciler.go

```diff
@@ -0,0 +1,205 @@
+)
+
+var _ SubResourceReconciler = (*KserveRawRouteReconciler)(nil)
+
+type KserveRawRouteReconciler struct {
+	client         client.Client
+	routeHandler   resources.RouteHandler
+	deltaProcessor processors.DeltaProcessor
+}
+
+func NewKserveRawRouteReconciler(client client.Client) *KserveRawRouteReconciler {
+	return &KserveRawRouteReconciler{
+		client:         client,
+		routeHandler:   resources.NewRouteHandler(client),
+		deltaProcessor: processors.NewDeltaProcessor(),
+	}
+}
+
+func (r *KserveRawRouteReconciler) Reconcile(ctx context.Context, log logr.Logger, isvc *kservev1beta1.InferenceService) error {
+	log.V(1).Info("Reconciling Generic Route for Kserve InferenceService")
+
+	// Create Desired resource
+	desiredResource, err := r.createDesiredResource(ctx, log, isvc)
+	if err != nil {
+		return err
+	}
+
+	// Get Existing resource
+	existingResource, err := r.getExistingResource(ctx, log, isvc)
+	if err != nil {
+		return err
+	}
+
+	// Process Delta
+	if err = r.processDelta(ctx, log, desiredResource, existingResource); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (r *KserveRawRouteReconciler) Delete(ctx context.Context, log logr.Logger, isvc *kservev1beta1.InferenceService) error {
+	log.V(1).Info("Deleting Kserve inference service generic route")
+	return r.routeHandler.DeleteRoute(ctx, types.NamespacedName{Name: isvc.Name, Namespace: isvc.Namespace})
+}
+
+func (r *KserveRawRouteReconciler) Cleanup(_ context.Context, _ logr.Logger, _ string) error {
+	// NOOP - resources are deleted together with ISVCs
+	return nil
+}
+
+func (r *KserveRawRouteReconciler) createDesiredResource(ctx context.Context, log logr.Logger, isvc *kservev1beta1.InferenceService) (*v1.Route, error) {
+	var err error
+	enableAuth := false
+	if enableAuth, err = strconv.ParseBool(isvc.Labels[constants.LabelEnableAuthODH]); err != nil {
+		enableAuth = false
+	}
+	createRoute := false
+	if val, ok := isvc.Labels[constants.KserveNetworkVisibility]; ok && val == constants.LabelEnableKserveRawRoute {
+		createRoute = true
+	}
+	if !createRoute {
+		log.Info("InferenceService does not have label '" + constants.KserveNetworkVisibility + "' annotation" +
+			" set to '" + constants.LabelEnableKserveRawRoute + "'. Skipping route creation")
+		return nil, nil
+	}
+
+	// Fetch the service with the label "serving.kserve.io/inferenceservice=isvc.Name" in the isvc namespace
+	serviceList := &corev1.ServiceList{}
+	labelSelector := client.MatchingLabels{constants.KserveGroupAnnotation: isvc.Name}
+	err = r.client.List(ctx, serviceList, client.InNamespace(isvc.Namespace), labelSelector)
+	if err != nil || len(serviceList.Items) == 0 {
+		log.Error(err, "Failed to fetch service for InferenceService", "InferenceService", isvc.Name)
+		return nil, err
+	}
+	var targetServiceName string
+	var servicePort int32
+	for _, service := range serviceList.Items {
+		if val, ok := service.Labels["component"]; ok && val == "predictor" {
```

Comment:
Similarly to [a comment in the other PR](https://github.com/opendatahub-io/kserve/pull/419#discussion_r1805380869), you cannot hard-code to the `predictor` service, because the Route would be bypassing the `transformer` in case the user has chosen to configure a transformer.

In serverless there is a _top-level_ service that would hide the detail of choosing between predictor and transformer. I'm not sure if Raw has it. If it exists, you should use that top-level service.

Reply:
raw doesn't have a top level service from what I saw, I added logic to check for the presence of a transformer service here, and if a transformer is detected that will be used as the target service for creating the route